File: packages/clinical-contracts/src/rule4/evaluator.ts
import { rule4EmptyResultFingerprintV1Hash } from './emptyResultFingerprintV1.js';
import { rule4Fingerprint } from './fingerprint.js';

import type { Rule4InputContract } from './input.js';
import {
  isRule4Phase2SafetyContract,
  toPhase2VerifiedAge,
  validateRule4InputContract,
} from './input.js';
import type { Rule4Result, Rule4SafetyGateOutput, Rule4SlotResult } from './output.js';
import { validateRule4OutputCodes } from './outputCodeValidation.js';
import { evaluateRule4SafetyGate } from './safety/evaluateSafetyGate.js';
import { evaluateEvidenceAdapter } from './evidence/evaluateEvidenceAdapter.js';
import type { Rule4EvidenceAdapterOutput } from './evidence/types.js';
import { evaluatePolarityAdapter } from './polarity/evaluatePolarityAdapter.js';
import type { Rule4PolarityAdapterOutput } from './polarity/types.js';
import { evaluatePhaseAdapter } from './phase/evaluatePhaseAdapter.js';
import type { Rule4PhaseAdapterOutput } from './phase/types.js';
import { evaluateSeverityAdapter } from './severity/evaluateSeverityAdapter.js';
import type { Rule4SeverityAdapterOutput } from './severity/types.js';
import { evaluateEligibilityAdapter } from './eligibility/evaluateEligibilityAdapter.js';
import type { Rule4EligibilityAdapterOutput } from './eligibility/types.js';
import { evaluateSelectionAdapter } from './selection/evaluateSelectionAdapter.js';
import type { Rule4SelectionAdapterOutput } from './selection/types.js';
import { evaluatePediatricOverlayAdapter } from './pediatricOverlay/evaluatePediatricOverlayAdapter.js';
import type { Rule4PediatricOverlayAdapterOutput } from './pediatricOverlay/types.js';
import { evaluateDoctorReviewAdapter } from './doctorReview/evaluateDoctorReviewAdapter.js';
import type { Rule4DoctorReviewAdapterOutput } from './doctorReview/types.js';
import {
  RULE4_CONTRACT_VERSION,
  RULE4_CONTRACT_VERSION_PHASE2,
  RULE4_DEFAULT_ENGINE_MODE,
  type Rule4EngineMode,
  type Rule4PotencyStatus,
} from './version.js';

function stableJson(value: unknown): string {
  return JSON.stringify(value, (_k, v) => (v === undefined ? null : v));
}

/** Phase 1 only — input serialization is not canonical cross-language. */
function fingerprintPhase1Result(body: Record<string, unknown>): string {
  return rule4Fingerprint(body);
}

export function parseRule4EngineMode(raw: string | null | undefined): Rule4EngineMode {
  if (raw === undefined || raw === null || raw.trim() === '') {
    return RULE4_DEFAULT_ENGINE_MODE;
  }
  const v = raw.trim().toLowerCase();
  if (v === 'off' || v === 'shadow' || v === 'active') {
    return v;
  }
  throw new Error('RULE4_ENGINE_MODE_INVALID');
}

function mapSafetyGateOutput(
  safety: ReturnType<typeof evaluateRule4SafetyGate>,
): Rule4SafetyGateOutput {
  return {
    safetyGateStatus: safety.safetyGateStatus,
    safetyStatus: safety.safetyStatus,
    prescriptionStatus: safety.prescriptionStatus,
    holdStatus: safety.holdStatus,
    patientWideHold: safety.patientWideHold,
    urgentEscalationRequired: safety.urgentEscalationRequired,
    analysisStatus: safety.analysisStatus,
    clinicalPrescriptionSummary: safety.clinicalPrescriptionSummary,
    d13HardStopActive: safety.d13HardStopActive,
    safetyNoticeKey: safety.safetyNoticeKey,
    safetyClearForFutureCascade: safety.safetyClearForFutureCascade,
    deterministicSafetyFingerprint: safety.deterministicSafetyFingerprint,
    reasonCodes: safety.reasonCodes,
    limitationCodes: safety.limitationCodes,
    ageVerificationStatus: safety.ageResolution.verificationStatus,
    pediatricBand: safety.ageResolution.pediatricBand,
  };
}

function slotOutcomeFromSafety(
  safety: ReturnType<typeof evaluateRule4SafetyGate>,
): Pick<Rule4SlotResult, 'potencyStatus' | 'reasonCodes' | 'limitationCodes'> {
  if (safety.patientWideHold || safety.d13HardStopActive) {
    return {
      potencyStatus: 'BLOCKED_BY_PATIENT_WIDE_SAFETY_GATE',
      reasonCodes: ['BLOCKED_BY_PATIENT_WIDE_SAFETY_GATE', ...safety.reasonCodes],
      limitationCodes: safety.limitationCodes,
    };
  }
  if (
    safety.ageResolution.verificationStatus === 'MISSING' ||
    safety.ageResolution.verificationStatus === 'INVALID' ||
    safety.ageResolution.verificationStatus === 'CONTRADICTORY' ||
    safety.ageResolution.verificationStatus === 'UNRESOLVED'
  ) {
    return {
      potencyStatus: 'UNRESOLVED',
      reasonCodes: [...new Set([...safety.ageResolution.reasonCodes, ...safety.reasonCodes])],
      limitationCodes: safety.limitationCodes,
    };
  }
  if (safety.safetyClearForFutureCascade) {
    return {
      potencyStatus: 'NOT_EVALUATED',
      reasonCodes: ['RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED', ...safety.reasonCodes],
      limitationCodes: ['PHASE1_NO_CLINICAL_EVALUATION', ...safety.limitationCodes],
    };
  }
  return {
    potencyStatus: 'UNRESOLVED',
    reasonCodes: [...new Set(safety.reasonCodes)],
    limitationCodes: safety.limitationCodes,
  };
}

function buildPhase1EmptySlots(input: Rule4InputContract): Rule4SlotResult[] {
  return input.formulaSlots.map((slot) => ({
    formulaSlotId: slot.formulaSlotId,
    formulaTargetId: slot.formulaTargetId,
    potencyStatus: 'NOT_EVALUATED' as const,
    selectedDilution: null,
    selectedCascade: null,
    reasonCodes: ['RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED'] as const,
    limitationCodes: ['PHASE1_NO_CLINICAL_EVALUATION'] as const,
    evidenceItemIds: [...slot.structuredEvidenceItemIds],
  }));
}

function evaluatePhase2Safety(input: Rule4InputContract): Rule4Result {
  const phase2Input = isRule4Phase2SafetyContract(input)
    ? input
    : {
        ...input,
        contractVersion: RULE4_CONTRACT_VERSION_PHASE2,
        verifiedAge: toPhase2VerifiedAge(input.verifiedAge),
      };

  const safety = evaluateRule4SafetyGate({
    ...phase2Input,
    contractVersion: RULE4_CONTRACT_VERSION_PHASE2,
    verifiedAge: toPhase2VerifiedAge(phase2Input.verifiedAge),
  });

  const slotTemplate = slotOutcomeFromSafety(safety);
  const slots: Rule4SlotResult[] = input.formulaSlots.map((slot) => ({
    formulaSlotId: slot.formulaSlotId,
    formulaTargetId: slot.formulaTargetId,
    potencyStatus: slotTemplate.potencyStatus as Rule4PotencyStatus,
    selectedDilution: null,
    selectedCascade: null,
    reasonCodes: [...new Set(slotTemplate.reasonCodes)],
    limitationCodes: [...new Set(slotTemplate.limitationCodes)],
    evidenceItemIds: [...slot.structuredEvidenceItemIds],
  }));

  const topReason = [...new Set([...safety.reasonCodes, ...slotTemplate.reasonCodes])];
  const topLimitation = [...new Set(safety.limitationCodes)];

  const core = {
    contractVersion: input.contractVersion,
    rulesetVersion: input.rulesetVersion,
    executionStatus: 'NOT_IMPLEMENTED' as const,
    engineMode: input.engineMode,
    automaticPotencyRuntime: false as const,
    automaticPrescriptionIssuanceRuntime: false as const,
    currentRuntimePotencyDelta: 'NONE' as const,
    currentRuntimeIssuanceDelta: 'NONE' as const,
    finalDoctorApprovalRequired: true as const,
    prescriptionIssueAllowed: false as const,
    slots,
    reasonCodes: topReason,
    limitationCodes: topLimitation,
    safetyGate: mapSafetyGateOutput(safety),
  };

  validateRule4OutputCodes(core);

  return {
    ...core,
    deterministicFingerprint: rule4EmptyResultFingerprintV1Hash({
      contractVersion: input.contractVersion,
      rulesetVersion: input.rulesetVersion,
      executionStatus: 'NOT_IMPLEMENTED',
      engineMode: input.engineMode,
      prescriptionIssueAllowed: false,
      deterministicSafetyFingerprint: safety.deterministicSafetyFingerprint,
      slots: slots.map((s) => ({
        formulaSlotId: s.formulaSlotId,
        potencyStatus: s.potencyStatus,
        reasonCodes: s.reasonCodes,
        limitationCodes: s.limitationCodes,
      })),
    }),
  };
}

/** Phase 1 empty evaluator — unchanged for phase-1 contract without safety extension. */
export function evaluateRule4Empty(input: Rule4InputContract): Rule4Result {
  validateRule4InputContract(input);
  if (input.engineMode === 'active') {
    throw new Error('RULE4_ENGINE_MODE_ACTIVE_NOT_IMPLEMENTED');
  }

  if (
    input.contractVersion === RULE4_CONTRACT_VERSION_PHASE2 ||
    input.bpReadings != null ||
    input.structuredCriticalFindings != null ||
    input.structuredFrozenRedFlags != null ||
    input.rawLabKeywordPresent != null
  ) {
    return evaluatePhase2Safety(input);
  }

  const slots = buildPhase1EmptySlots(input);

  const core = {
    contractVersion: RULE4_CONTRACT_VERSION,
    rulesetVersion: input.rulesetVersion,
    executionStatus: 'NOT_IMPLEMENTED' as const,
    engineMode: input.engineMode,
    automaticPotencyRuntime: false as const,
    automaticPrescriptionIssuanceRuntime: false as const,
    currentRuntimePotencyDelta: 'NONE' as const,
    currentRuntimeIssuanceDelta: 'NONE' as const,
    finalDoctorApprovalRequired: true as const,
    prescriptionIssueAllowed: false as const,
    slots,
    reasonCodes: ['RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED'] as const,
    limitationCodes: ['PHASE1_NO_CLINICAL_EVALUATION'] as const,
  };

  validateRule4OutputCodes(core);

  return {
    ...core,
    deterministicFingerprint: fingerprintPhase1Result({
      ...core,
      inputFingerprint: stableJson(input),
    }),
  };
}

export function evaluateRule4Phase2Safety(input: Rule4InputContract): Rule4Result {
  validateRule4InputContract(input);
  if (input.engineMode === 'active') {
    throw new Error('RULE4_ENGINE_MODE_ACTIVE_NOT_IMPLEMENTED');
  }
  return evaluatePhase2Safety(input);
}

export type Rule4ShadowEvaluationBundle = {
  result: Rule4Result;
  evidenceAdapter: Rule4EvidenceAdapterOutput | null;
  polarityRouting: Rule4PolarityAdapterOutput | null;
  phaseResolution: Rule4PhaseAdapterOutput | null;
  severityResolution: Rule4SeverityAdapterOutput | null;
  eligibilityResolution: Rule4EligibilityAdapterOutput | null;
  selectionResolution: Rule4SelectionAdapterOutput | null;
  pediatricOverlayResolution: Rule4PediatricOverlayAdapterOutput | null;
  doctorReviewResolution: Rule4DoctorReviewAdapterOutput | null;
};

/**
 * Phase 2 public result unchanged; Phase 3 evidence envelope computed only in shadow when provided.
 */
export function evaluateRule4ShadowBundle(input: Rule4InputContract): Rule4ShadowEvaluationBundle {
  const result = evaluateRule4Empty(input);
  let evidenceAdapter: Rule4EvidenceAdapterOutput | null = null;
  let polarityRouting: Rule4PolarityAdapterOutput | null = null;
  let phaseResolution: Rule4PhaseAdapterOutput | null = null;
  let severityResolution: Rule4SeverityAdapterOutput | null = null;
  let eligibilityResolution: Rule4EligibilityAdapterOutput | null = null;
  let selectionResolution: Rule4SelectionAdapterOutput | null = null;
  let pediatricOverlayResolution: Rule4PediatricOverlayAdapterOutput | null = null;
  let doctorReviewResolution: Rule4DoctorReviewAdapterOutput | null = null;

  if (input.engineMode === 'shadow') {
    if (input.evidenceAdapter) {
      evidenceAdapter = evaluateEvidenceAdapter(input.evidenceAdapter);
    }
    if (input.polarityAdapter) {
      polarityRouting = evaluatePolarityAdapter(input.polarityAdapter, {
        safetyGate: result.safetyGate ?? null,
        evidenceAdapter,
        bindingGateMandatory: true,
      });
    }
    if (input.phaseAdapter) {
      phaseResolution = evaluatePhaseAdapter(input.phaseAdapter, {
        safetyGate: result.safetyGate ?? null,
        evidenceAdapter,
        polarityRouting,
        bindingGateMandatory: true,
      });
    }
    if (input.severityAdapter) {
      severityResolution = evaluateSeverityAdapter(input.severityAdapter, {
        safetyGate: result.safetyGate ?? null,
        evidenceAdapter,
        polarityRouting,
        phaseResolution,
        bindingGateMandatory: true,
      });
    }
    if (input.eligibilityAdapter) {
      eligibilityResolution = evaluateEligibilityAdapter(input.eligibilityAdapter, {
        safetyGate: result.safetyGate ?? null,
        evidenceAdapter,
        polarityRouting,
        phaseResolution,
        severityResolution,
        bindingGateMandatory: true,
      });
    }
    if (input.selectionAdapter) {
      selectionResolution = evaluateSelectionAdapter(input.selectionAdapter, {
        safetyGate: result.safetyGate ?? null,
        verifiedAge: input.verifiedAge,
        eligibilityResolution,
        evidenceAdapter: evidenceAdapter,
        evidenceItems: input.evidenceAdapter?.items ?? null,
        bindingGateMandatory: true,
      });
    }
    if (input.pediatricOverlayAdapter && selectionResolution) {
      pediatricOverlayResolution = evaluatePediatricOverlayAdapter(input.pediatricOverlayAdapter, {
        safetyGate: result.safetyGate ?? null,
        verifiedAge: input.verifiedAge,
        selectionResolution,
        upstreamEligibilityFingerprint:
          eligibilityResolution?.deterministicCandidateEligibilityFingerprint ?? null,
      });
    }
    if (input.doctorReviewAdapter) {
      doctorReviewResolution = evaluateDoctorReviewAdapter(
        input.doctorReviewAdapter,
        input.doctorReviewContext ?? {},
      );
    }
  }

  return {
    result,
    evidenceAdapter,
    polarityRouting,
    phaseResolution,
    severityResolution,
    eligibilityResolution,
    selectionResolution,
    pediatricOverlayResolution,
    doctorReviewResolution,
  };
}

File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/evaluator.py
from __future__ import annotations

import copy

from ..fingerprints import fingerprint
from .empty_result_fingerprint_v1 import rule4_empty_result_fingerprint_v1_hash
from .registry_loader import load_reason_code_registry
from .registry_validation import validate_rule4_output_codes
from .safety.evaluate_safety_gate import evaluate_safety_gate
from .validate import RULE4_CONTRACT_VERSION, RULE4_CONTRACT_VERSION_PHASE2, validate_rule4_input_contract


class Rule4ConfigurationError(ValueError):
    code = "RULE4_CONFIGURATION_ERROR"


__all__ = [
    "Rule4ConfigurationError",
    "evaluate_rule4_empty",
    "load_reason_code_registry",
]


def _uses_phase2_safety(input_contract: dict) -> bool:
    if input_contract.get("contract_version") == RULE4_CONTRACT_VERSION_PHASE2:
        return True
    if input_contract.get("bp_readings") is not None:
        return True
    if input_contract.get("structured_critical_findings") is not None:
        return True
    if input_contract.get("structured_frozen_red_flags") is not None:
        return True
    if input_contract.get("raw_lab_keyword_present") is not None:
        return True
    return False


def _map_safety_gate(safety) -> dict:
    return {
        "safety_gate_status": safety.safety_gate_status,
        "safety_status": safety.safety_status,
        "prescription_status": safety.prescription_status,
        "hold_status": safety.hold_status,
        "patient_wide_hold": safety.patient_wide_hold,
        "urgent_escalation_required": safety.urgent_escalation_required,
        "analysis_status": safety.analysis_status,
        "clinical_prescription_summary": safety.clinical_prescription_summary,
        "d13_hard_stop_active": safety.d13_hard_stop_active,
        "safety_notice_key": safety.safety_notice_key,
        "safety_clear_for_future_cascade": safety.safety_clear_for_future_cascade,
        "deterministic_safety_fingerprint": safety.deterministic_safety_fingerprint,
        "reason_codes": safety.reason_codes,
        "limitation_codes": safety.limitation_codes,
        "age_verification_status": safety.age_resolution.verification_status,
        "pediatric_band": safety.age_resolution.pediatric_band,
    }


def _slot_template_from_safety(safety) -> dict:
    if safety.patient_wide_hold or safety.d13_hard_stop_active:
        return {
            "potency_status": "BLOCKED_BY_PATIENT_WIDE_SAFETY_GATE",
            "reason_codes": sorted(set(["BLOCKED_BY_PATIENT_WIDE_SAFETY_GATE", *safety.reason_codes])),
            "limitation_codes": safety.limitation_codes,
        }
    age = safety.age_resolution
    if age.verification_status in {"MISSING", "INVALID", "CONTRADICTORY", "UNRESOLVED"}:
        return {
            "potency_status": "UNRESOLVED",
            "reason_codes": sorted(set([*age.reason_codes, *safety.reason_codes])),
            "limitation_codes": safety.limitation_codes,
        }
    if safety.safety_clear_for_future_cascade:
        return {
            "potency_status": "NOT_EVALUATED",
            "reason_codes": sorted(
                set(["RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED", *safety.reason_codes])
            ),
            "limitation_codes": sorted(
                set(["PHASE1_NO_CLINICAL_EVALUATION", *safety.limitation_codes])
            ),
        }
    return {
        "potency_status": "UNRESOLVED",
        "reason_codes": sorted(set(safety.reason_codes)),
        "limitation_codes": safety.limitation_codes,
    }


def _evaluate_phase2(input_contract: dict) -> dict:
    safety = evaluate_safety_gate(input_contract, fingerprint_fn=fingerprint)
    slot_tpl = _slot_template_from_safety(safety)
    slots_out = []
    for slot in input_contract.get("formula_slots") or []:
        slots_out.append(
            {
                "formula_slot_id": slot["formula_slot_id"],
                "formula_target_id": slot.get("formula_target_id"),
                "potency_status": slot_tpl["potency_status"],
                "selected_dilution": None,
                "selected_cascade": None,
                "reason_codes": slot_tpl["reason_codes"],
                "limitation_codes": slot_tpl["limitation_codes"],
                "evidence_item_ids": list(slot.get("structured_evidence_item_ids") or []),
            }
        )
    core = {
        "contract_version": input_contract.get("contract_version", RULE4_CONTRACT_VERSION_PHASE2),
        "ruleset_version": input_contract["ruleset_version"],
        "execution_status": "NOT_IMPLEMENTED",
        "engine_mode": input_contract.get("engine_mode", "shadow"),
        "automatic_potency_runtime": False,
        "automatic_prescription_issuance_runtime": False,
        "current_runtime_potency_delta": "NONE",
        "current_runtime_issuance_delta": "NONE",
        "final_doctor_approval_required": True,
        "prescription_issue_allowed": False,
        "slots": slots_out,
        "reason_codes": slot_tpl["reason_codes"],
        "limitation_codes": safety.limitation_codes,
        "safety_gate": _map_safety_gate(safety),
    }
    validate_rule4_output_codes(core)
    core["deterministic_fingerprint"] = rule4_empty_result_fingerprint_v1_hash(
        contract_version=core["contract_version"],
        ruleset_version=core["ruleset_version"],
        execution_status=core["execution_status"],
        engine_mode=core["engine_mode"],
        prescription_issue_allowed=core["prescription_issue_allowed"],
        deterministic_safety_fingerprint=safety.deterministic_safety_fingerprint,
        slots=slots_out,
    )
    return core


def evaluate_rule4_shadow_bundle(input_contract: dict) -> dict:
    result = evaluate_rule4_empty(input_contract)
    evidence = None
    polarity = None
    phase_resolution = None
    severity_resolution = None
    eligibility_resolution = None
    selection_resolution = None
    pediatric_overlay_resolution = None
    doctor_review_resolution = None
    if input_contract.get("engine_mode") == "shadow":
        if input_contract.get("evidence_adapter"):
            from .evidence.evaluate_evidence_adapter import evaluate_evidence_adapter

            evidence = evaluate_evidence_adapter(input_contract["evidence_adapter"])
        if input_contract.get("polarity_adapter"):
            from .polarity.evaluate_polarity_adapter import evaluate_polarity_adapter

            polarity = evaluate_polarity_adapter(
                input_contract["polarity_adapter"],
                safety_gate=result.get("safety_gate"),
                evidence_adapter=evidence,
                binding_gate_mandatory=True,
            )
        if input_contract.get("phase_adapter"):
            from .phase.evaluate_phase_adapter import evaluate_phase_adapter

            phase_resolution = evaluate_phase_adapter(
                input_contract["phase_adapter"],
                safety_gate=result.get("safety_gate"),
                evidence_adapter=evidence,
                polarity_routing=polarity,
                binding_gate_mandatory=True,
            )
        severity_resolution = None
        if input_contract.get("severity_adapter"):
            from .severity.evaluate_severity_adapter import evaluate_severity_adapter

            severity_resolution = evaluate_severity_adapter(
                input_contract["severity_adapter"],
                safety_gate=result.get("safety_gate"),
                evidence_adapter=evidence,
                polarity_routing=polarity,
                phase_resolution=phase_resolution,
                binding_gate_mandatory=True,
            )
        if input_contract.get("eligibility_adapter"):
            from .eligibility.evaluate_eligibility_adapter import evaluate_eligibility_adapter

            eligibility_resolution = evaluate_eligibility_adapter(
                input_contract["eligibility_adapter"],
                safety_gate=result.get("safety_gate"),
                evidence_adapter=evidence,
                polarity_routing=polarity,
                phase_resolution=phase_resolution,
                severity_resolution=severity_resolution,
                binding_gate_mandatory=True,
            )
        if input_contract.get("selection_adapter"):
            from .selection.evaluate_selection_adapter import evaluate_selection_adapter

            evidence_items = None
            if input_contract.get("evidence_adapter"):
                evidence_items = input_contract["evidence_adapter"].get("items")
            selection_resolution = evaluate_selection_adapter(
                input_contract["selection_adapter"],
                {
                    "safety_gate": result.get("safety_gate"),
                    "verified_age": input_contract.get("verified_age"),
                    "eligibility_resolution": eligibility_resolution,
                    "evidence_adapter": evidence,
                    "evidence_items": evidence_items,
                    "binding_gate_mandatory": True,
                },
            )
        if input_contract.get("pediatric_overlay_adapter") and selection_resolution:
            from .pediatric_overlay.evaluate_pediatric_overlay_adapter import (
                evaluate_pediatric_overlay_adapter,
            )

            upstream_fp = None
            if eligibility_resolution:
                upstream_fp = eligibility_resolution.get(
                    "deterministic_candidate_eligibility_fingerprint"
                )
            pediatric_overlay_resolution = evaluate_pediatric_overlay_adapter(
                input_contract["pediatric_overlay_adapter"],
                {
                    "safety_gate": result.get("safety_gate"),
                    "verified_age": input_contract.get("verified_age"),
                    "selection_resolution": selection_resolution,
                    "upstream_eligibility_fingerprint": upstream_fp,
                },
            )
        if input_contract.get("doctor_review_adapter"):
            from .doctor_review.evaluate_doctor_review_adapter import evaluate_doctor_review_adapter

            doctor_review_resolution = evaluate_doctor_review_adapter(
                input_contract["doctor_review_adapter"],
                input_contract.get("doctor_review_context") or {},
            )
    return {
        "result": result,
        "evidence_adapter": evidence,
        "polarity_routing": polarity,
        "phase_resolution": phase_resolution,
        "severity_resolution": severity_resolution,
        "eligibility_resolution": eligibility_resolution,
        "selection_resolution": selection_resolution,
        "pediatric_overlay_resolution": pediatric_overlay_resolution,
        "doctor_review_resolution": doctor_review_resolution,
    }


def evaluate_rule4_empty(input_contract: dict) -> dict:
    validate_rule4_input_contract(input_contract)
    mode = input_contract.get("engine_mode", "off")
    if mode == "active":
        raise Rule4ConfigurationError("RULE4_ENGINE_MODE_ACTIVE_NOT_IMPLEMENTED")

    if _uses_phase2_safety(input_contract):
        return _evaluate_phase2(input_contract)

    slots_out = []
    for slot in input_contract.get("formula_slots") or []:
        slots_out.append(
            {
                "formula_slot_id": slot["formula_slot_id"],
                "formula_target_id": slot.get("formula_target_id"),
                "potency_status": "NOT_EVALUATED",
                "selected_dilution": None,
                "selected_cascade": None,
                "reason_codes": ["RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED"],
                "limitation_codes": ["PHASE1_NO_CLINICAL_EVALUATION"],
                "evidence_item_ids": list(slot.get("structured_evidence_item_ids") or []),
            }
        )

    core = {
        "contract_version": RULE4_CONTRACT_VERSION,
        "ruleset_version": input_contract["ruleset_version"],
        "execution_status": "NOT_IMPLEMENTED",
        "engine_mode": mode,
        "automatic_potency_runtime": False,
        "automatic_prescription_issuance_runtime": False,
        "current_runtime_potency_delta": "NONE",
        "current_runtime_issuance_delta": "NONE",
        "final_doctor_approval_required": True,
        "prescription_issue_allowed": False,
        "slots": slots_out,
        "reason_codes": ["RULE4_PHASE1_EVALUATOR_NOT_IMPLEMENTED"],
        "limitation_codes": ["PHASE1_NO_CLINICAL_EVALUATION"],
    }
    validate_rule4_output_codes(core)
    fp_body = {**core, "input_fingerprint": fingerprint(input_contract)}
    core["deterministic_fingerprint"] = fingerprint(fp_body)
    return core
